### PR TITLE
Tekton Task acceptance tests

### PR DESCRIPTION
**Pull request by zregvart**
Implements acceptance tests for the verify-enterprise-contract Tekton Task. The acceptance test harness launches a Kind cluster, installs from Kustomization from `hack/test` the required components: image registry, Tekton Pipelines & Webhooks, EnterpriseContractPolicy custom resource.

The current implementations of feature steps allow for creation of a working namespace provisioned via `hack/work` Kustomization, EnterpriseContractPolicy object creation, launching the verify-enterprise-contract Tekton Task via TaskRun, and asserting that the Task succeeded.

The ec-cli and the verify-enterprise-contract Tekton Task bundle images are built and pushed to the cluster. The Tekton Task bundle image is tagged with the version from the directory it is placed in, e.g.:

    registry.image-registry.svc.cluster.local:46859/ec-task-bundle:0.1

And the ec-cli image is tagged with the `latest-<os>-<arch>`, e.g.:

    registry.image-registry.svc.cluster.local:46859/ec-cli:latest-linux-amd64

Ref. https://issues.redhat.com/browse/HACBS-1420

### CODE REVIEW

**Comment by zregvart** [comment]
In draft, the current to-do list is:

- [ ] build ec image and v-e-c task bundle, push them to the cluster registry
- [ ] make sure Kind cluster's name is unique (seed the random)
- [ ] better diagnostics on task failure
- [ ] code comments

**Comment by zregvart** [comment]
The acceptance test now runs the Task with `STRICT=true` and using a policy with the single rule (`slsa_provenance_available`). A new commit was added needed to fix an issue when the policy includes the public key:

[Remove default for PUBLIC_KEY parameter](https://github.com/hacbs-contract/ec-cli/pull/306/commits/f3085d8c8d3651d00a3c61c3edf095462b1b1106)

Given that the policy itself can provide the public key needed for signature verification we should not have a default for the `PUBLIC_KEY` parameter. Having the default there will override the public key from the policy and the signature would be deemed invalid.

**Review by lcarva — APPROVED** [review]
This is great!

`make acceptance` takes `1m43.192s` on my system which is quite alright given what is being covered.

:rocket: :heavy_check_mark:

**Comment by simonbaird** [comment]
Running this:
```
go test -v -tags=acceptance ./internal/acceptance -args -persist -tags=@focus
```

It worked nicely.

Then I ran it again and hit this failures:

```
--- Failed steps:

  Scenario: Golden container image # /home/sbaird/dev/ec-cli/features/task.feature:8
    Given a cluster running # /home/sbaird/dev/ec-cli/features/task.feature:6
      Error: could not find a log line that matches "Reached target .*Multi-User System.*|detected cgroup v1"
```

**Comment by simonbaird** [comment]
If I use `make acceptance` it passes and doesn't leave the kind controller container running, so I can run that multiple times and get multiple passes.

The time taken on my system is around 85 seconds.

**Review by simonbaird — APPROVED** [review]
Lgtm, but consider if there's an easy fix for the left over kind controllers I saw when running the individual feature.

File: internal/acceptance/kubernetes/kind/image.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	return nil
+}
+
+// buildTaskBundleImage runs `make task-bundle` for each version of the Task in
+// the `$REPOSITORY_ROOT/task` directory to push the Tekton Task bundle to the
+// registry runing on the Kind cluster. The image is pushed to image reference:
+// `localhost:<registry-port>/ec-task-bundle:<version>`, so each bundle contains
+// only the task of a particular version. The image reference to the ec-cli
+// image is replaced with the image reference from buildCliImage.
+func (k *kindCluster) buildTaskBundleImage(ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (zregvart) [reply internal/acceptance/kubernetes/kind/image.go:56]: Make target would be a good addition indeed, I've created [HACBS-1520](https://issues.redhat.com/browse/HACBS-1520)

**Comment by zregvart** [comment]
I managed to get the same error Simon got above, from docker daemon logs:

```
Dec 19 19:09:45 zregvart dockerd[69135]: time="2022-12-19T19:09:45.403804715+01:00" level=info msg="No non-localhost DNS nameservers are left in resolv.conf. Using default external servers: [nameserver 8.8.8.8 nameserver 8.8.4.4]"
Dec 19 19:09:45 zregvart dockerd[69135]: time="2022-12-19T19:09:45.404105415+01:00" level=info msg="IPv6 enabled; Adding default IPv6 external servers: [nameserver 2001:4860:4860::8888 nameserver 2001:4860:4860::8844]"
Dec 19 19:09:45 zregvart dockerd[69165]: time="2022-12-19T19:09:45.474304799+01:00" level=info msg="loading plugin \"io.containerd.event.v1.publisher\"..." runtime=io.containerd.runc.v2 type=io.containerd.event.v1
Dec 19 19:09:45 zregvart dockerd[69165]: time="2022-12-19T19:09:45.474491735+01:00" level=info msg="loading plugin \"io.containerd.internal.v1.shutdown\"..." runtime=io.containerd.runc.v2 type=io.containerd.internal.v1
Dec 19 19:09:45 zregvart dockerd[69165]: time="2022-12-19T19:09:45.474506949+01:00" level=info msg="loading plugin \"io.containerd.ttrpc.v1.task\"..." runtime=io.containerd.runc.v2 type=io.containerd.ttrpc.v1
Dec 19 19:09:45 zregvart dockerd[69165]: time="2022-12-19T19:09:45.474636374+01:00" level=info msg="starting signal loop" namespace=moby path=/run/docker/containerd/daemon/io.containerd.runtime.v2.task/moby/d122277e9252f8d00dc3bfd2a162ee91dfabda11b76bf110d2bdacb1cef1f089 pid=329338 runtime=io.containerd.runc.v2
Dec 19 19:09:45 zregvart dockerd[329387]: time="2022-12-19T19:09:45+01:00" level=info msg="Firewalld: docker zone already exists, returning"
Dec 19 19:09:45 zregvart dockerd[69165]: time="2022-12-19T19:09:45.709065334+01:00" level=warning msg="error from *cgroupsv2.Manager.EventChan" error="failed to create inotify fd"
Dec 19 19:09:45 zregvart d122277e9252[69135]: INFO: ensuring we can execute mount/umount even with userns-remap
Dec 19 19:09:45 zregvart d122277e9252[69135]: INFO: remounting /sys read-only
Dec 19 19:09:45 zregvart d122277e9252[69135]: INFO: making mounts shared
Dec 19 19:09:45 zregvart d122277e9252[69135]: INFO: detected cgroup v2
Dec 19 19:09:45 zregvart d122277e9252[69135]: INFO: clearing and regenerating /etc/machine-id
Dec 19 19:09:45 zregvart dockerd[69135]: time="2022-12-19T19:09:45.738792727+01:00" level=error msg="Error streaming logs: error getting journald fd: Too many open files" container=d122277e9252f8d00dc3bfd2a162ee91dfabda11b76bf110d2bdacb1cef1f089 method="(*Daemon).ContainerLogs" module=daemon
Dec 19 19:09:45 zregvart d122277e9252[69135]: Initializing machine ID from random generator.
Dec 19 19:09:45 zregvart d122277e9252[69135]: INFO: faking /sys/class/dmi/id/product_name to be "kind"
Dec 19 19:09:45 zregvart d122277e9252[69135]: INFO: faking /sys/class/dmi/id/product_uuid to be random
Dec 19 19:09:45 zregvart d122277e9252[69135]: INFO: faking /sys/devices/virtual/dmi/id/product_uuid as well
Dec 19 19:09:45 zregvart d122277e9252[69135]: INFO: setting iptables to detected mode: legacy
Dec 19 19:09:45 zregvart dockerd[69165]: time="2022-12-19T19:09:45.801405004+01:00" level=info msg="shim disconnected" id=d122277e9252f8d00dc3bfd2a162ee91dfabda11b76bf110d2bdacb1cef1f089
Dec 19 19:09:45 zregvart dockerd[69135]: time="2022-12-19T19:09:45.801415476+01:00" level=info msg="ignoring event" container=d122277e9252f8d00dc3bfd2a162ee91dfabda11b76bf110d2bdacb1cef1f089 module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Dec 19 19:09:45 zregvart dockerd[69165]: time="2022-12-19T19:09:45.801793447+01:00" level=warning msg="cleaning up after shim disconnected" id=d122277e9252f8d00dc3bfd2a162ee91dfabda11b76bf110d2bdacb1cef1f089 namespace=moby
Dec 19 19:09:45 zregvart dockerd[69165]: time="2022-12-19T19:09:45.801810961+01:00" level=info msg="cleaning up dead shim"
Dec 19 19:09:45 zregvart dockerd[69165]: time="2022-12-19T19:09:45.807821647+01:00" level=warning msg="cleanup warnings time=\"2022-12-19T19:09:45+01:00\" level=info msg=\"starting signal loop\" namespace=moby pid=329548 runtime=io.containerd.runc.v2\n"
```

I'm now getting `Failed to allocate directory watch: Too many open files` when I try systemctl restart. Odd...

**Comment by lcarva** [comment]
> I'm now getting `Failed to allocate directory watch: Too many open files` when I try systemctl restart. Odd...

Might have to bump the inotify values. Here's what I have:
```
$ sysctl fs.inotify
fs.inotify.max_queued_events = 16384
fs.inotify.max_user_instances = 512
fs.inotify.max_user_watches = 524288
```

**Comment by zregvart** [comment]
I'll merge this, created [HACBS-1523](https://issues.redhat.com/browse/HACBS-1523) to followup on `-persisted` issues.

**Comment by simonbaird** [comment]
Actually I didn't intentionally want it to -persist, when I was trying the focus test - I think just copy/pasted the command without paying enough attention.

**Pull request merged**